### PR TITLE
qemu.tests: add case to run qemu_iotests

### PR DESCRIPTION
**Pull request by pingl2016**
1. Update cases for qcow2/raw/vmdk image format
2. Add case for vhdx image format

Signed-off-by: Ping Li <pingl@redhat.com>

ID: 1395551, 1395552

### CODE REVIEW

File: qemu/tests/qemu_iotests.py [hunk 1]
```diff
@@ -1,50 +1,88 @@
...
+    version_pattern = params.get("qemu_version_pattern")
+    result_pattern = params.get("iotests_result_pattern")
+    lbranch = params.get("qemu_lbranch", None)
+    commit = params.get("qemu_commit", None)
+    base_uri = params.get("qemu_base_uri", None)
+    submodule = params.get("qemu_submodule")
+    destination_dir = params.get("code_destination_dir")
+
+    logging.info("Install packages for configure qemu source code")
+    yum_cmd = "rpm -q {0} || yum -y install {0}"
```
Comment (PyLearner) [inline qemu/tests/qemu_iotests.py]: is this function utils_misc.yum_install works for you ?

File: qemu/tests/qemu_iotests.py [hunk 2]
```diff
@@ -1,50 +1,86 @@
...
+    image_format = params.get("qemu_io_image_format")
+    version_pattern = params.get("qemu_version_pattern")
+    result_pattern = params.get("iotests_result_pattern")
+    lbranch = params.get("qemu_lbranch", None)
+    commit = params.get("qemu_commit", None)
+    base_uri = params.get("qemu_base_uri", None)
+    submodule = params.get("qemu_submodule")
+    destination_dir = params.get("code_destination_dir")
+
+    logging.info("Install packages for configure qemu source code")
```
Comment (CongLi) [inline qemu/tests/qemu_iotests.py]: context is better here.

File: qemu/tests/qemu_iotests.py [hunk 3]
```diff
@@ -1,50 +1,86 @@
...
+        cmd += " -%s" % image_format
+        output = process.system_output(cmd, ignore_status=True, shell=True)
+        match = re.search(result_pattern, output, re.I | re.M)
+        if match:
+            iotests_log_file = "qemu_iotests_%s.log" % image_format
+            iotests_log_file = utils_misc.get_path(test.debugdir, iotests_log_file)
+            file(iotests_log_file, "w+").write(output)
+            msg = "Total test %s cases, %s failed"
+            raise exceptions.TestFail(msg % (match.group(2), match.group(1)))
+    finally:
```
Comment (CongLi) [inline qemu/tests/qemu_iotests.py]: need ignore errors in finally, or it will cover the real errors

File: qemu/tests/cfg/qemu_iotests.cfg [hunk 4]
```diff
@@ -1,23 +1,23 @@
...
-    qemu_io_lbranch = master
-    #qemu_io_commit =
-    #qemu_io_base_uri =
-    #qemu_io_extra_options =
+    start_vm = no
+    kill_vm_before_test = yes
+    check_image = no
+    configure_deps_pkgs = zlib-devel gtk2-devel
+    qemu_submodule = dtc
+    qemu_url = "git://git.app.eng.bos.redhat.com/virt/rhel7/qemu-kvm.git"
```
Comment (CongLi) [inline qemu/tests/cfg/qemu_iotests.cfg]: Need confirm is this url suitable ?

**Comment by CongLi** [comment]
LGTM,
acked-by: Cong Li <coli@redhat.com>

**Comment by PyLearner** [comment]
LGTM,
Acked-by: Meng Yang meyang@redhat.com

**Comment by suqinhuang** [comment]
Command '/bin/git checkout qemu-kvm-rhev-2.6.0-28.el7.1' failed (rc=1)

File: qemu/tests/cfg/qemu_iotests.cfg [hunk 5]
```diff
@@ -1,27 +1,31 @@
...
+        need_run_configure = yes
+    qemu_version_pattern = "(?<=\()\S+(?=\))"
+    iotests_result_pattern = "Failed\s*(\d+)\s*of\s*(\d+)\s*test"
+    download_root_url = "http://download.xxx.com"
     variants:
         - raw_format:
-            only raw raw_dd
             qemu_io_image_format = raw
+            Host_RHEL.m7:
+                no Host_RHEL.m7.0, Host_RHEL.m7.1, Host_RHEL.m7.2
```
Comment (CongLi) [inline qemu/tests/cfg/qemu_iotests.cfg]: same as line 10, no need here.

File: qemu/tests/cfg/qemu_iotests.cfg [hunk 6]
```diff
@@ -1,27 +1,31 @@
...
-    qemu_io_lbranch = master
-    #qemu_io_commit =
-    #qemu_io_base_uri =
-    #qemu_io_extra_options =
+    start_vm = no
+    kill_vm_before_test = yes
+    check_image = no
+    configure_deps_pkgs = "rpm-build"
+    Host_RHEL.m7:
+        no Host_RHEL.m7.0, Host_RHEL.m7.1, Host_RHEL.m7.2
```
Comment (CongLi) [inline qemu/tests/cfg/qemu_iotests.cfg]: It should be Host_RHEL.m7.u0

Comment (CongLi) [inline qemu/tests/cfg/qemu_iotests.cfg]: It's better to add line 10 under line 2

File: qemu/tests/qemu_iotests.py [hunk 7]
```diff
@@ -1,50 +1,136 @@
...
+@error_context.context_aware
 def run(test, params, env):
     """
-    Fetch from git and run qemu-iotests using the qemu binaries under test.
+    Get qemu src code from src rpm and run qemu-iotests using the
+    qemu binaries.
 
-    1) Fetch qemu-io from git
+    1) Get url of src rpm from brew
+    2) Upack src code and apply patches
```
Comment (CongLi) [inline qemu/tests/qemu_iotests.py]: typo of "unpack", maybe decompress is better.

File: qemu/tests/cfg/qemu_iotests.cfg [hunk 8]
```diff
@@ -1,27 +1,31 @@
 - qemu_iotests:
+    only Linux
```
Comment (xutian) [inline qemu/tests/cfg/qemu_iotests.cfg]: no need this line, you have set only RHEL.m7

Comment (pingl2016) [inline qemu/tests/cfg/qemu_iotests.cfg]: I think we need the parameter. as "only Linux" is used for guest and " only Host_RHEL.m7, Host_RHEL.m6" is set for host

File: qemu/tests/qemu_iotests.py [hunk 9]
```diff
@@ -1,50 +1,156 @@
...
+        if match:
+            iotests_log_file = "qemu_iotests_%s.log" % image_format
+            iotests_log_file = utils_misc.get_path(test.debugdir, iotests_log_file)
+            file(iotests_log_file, "w+").write(output)
+            msg = "Total test %s cases, %s failed"
+            raise exceptions.TestFail(msg % (match.group(2), match.group(1)))
+
+    rpmbuild_dir = params.get("rpmbuild_dir", "/root/rpmbuild")
+    error_context.context("Install packages", logging.info)
+    if params.get("configure_deps_pkgs"):
```
Comment (CongLi) [inline qemu/tests/qemu_iotests.py]: params["configure_deps_pkgs"] if it's a must.

**Comment by CongLi** [comment]
Acked-by: Cong Li <coli@redhat.com>

**Comment by xutian** [comment]
@pingl2016  not a good idea to update old qemu-iotest. It will block -iotest with upstream qemu.  My suggest is refine a case named rh_qemu_iotest to run iotest in our src rpm package.

File: qemu/tests/rh_qemu_iotests.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+            retry -= 1
+            try:
+                return process.system_output(cmd, shell=True)
+            except process.CmdError, detail:
+                msg = "Fail to execute command"
+                logging.error("%s: %s." % (msg, detail))
+        raise exceptions.TestError("%s after %s times retry: %s" %
+                                   (msg, max_retry, detail))
+
+    def get_qemu_src_rpm_url():
```
Comment (xutian) [inline qemu/tests/rh_qemu_iotests.py]: Not need this function, please use brew download it directly.
srpm=$(rpm -qf  /usr/libexec/qemu-kvm|sed 's/x86_64/src/').rpm; brew dowload-build --rpm "$srpm"

can help you download srpm directly.

Comment (xutian) [inline qemu/tests/rh_qemu_iotests.py]: en, rpm has a --queryformat option
the command can be
`brew download-build --rpm $(rpm -qf `which qemu-kvm` --queryformat="%{NAME}-%{VERSION}.%{RELEASE}.src.rpm")`

File: qemu/tests/rh_qemu_iotests.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+            src_brew_link = re.findall(src_rpm_pattern, buildinfo, re.I)[0]
+        except IndexError:
+            raise exceptions.TestError("Could not get qemu src package "
+                                       "brew link matching pattern '%s'" %
+                                       src_rpm_pattern)
+        url = download_root + src_brew_link[11:]
+        logging.info("Brew URL is %s" % url)
+        return url
+
+    def get_qemu_src_from_rpm(rpmbuild_dir, url):
```
Comment (xutian) [inline qemu/tests/rh_qemu_iotests.py]: drop it

File: qemu/tests/rh_qemu_iotests.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+        cmd = "rpm -ivhf %s" % url
+        retry_command(cmd)
+        spec = os.path.join(rpmbuild_dir, 'SPECS/qemu-kvm.spec')
+        BUILD_dir = os.path.join(rpmbuild_dir, 'BUILD')
+        cmd = "rpmbuild -bp %s --nodeps" % spec
+        process.system_output(cmd, shell=True)
+        src_dir = os.listdir(BUILD_dir)[0]
+        return os.path.join(BUILD_dir, src_dir)
+
+    def run_configure_script(qemu_src_dir):
```
Comment (xutian) [inline qemu/tests/rh_qemu_iotests.py]: I guest it only need to do after a version, better to use package version as check condition

File: qemu/tests/rh_qemu_iotests.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+        cmd = './check'
+        if extra_options:
+            cmd += " %s" % extra_options
+        cmd += " -%s" % image_format
+        output = process.system_output(cmd, ignore_status=True, shell=True)
+        match = re.search(result_pattern, output, re.I | re.M)
+        if match:
+            iotests_log_file = "qemu_iotests_%s.log" % image_format
+            iotests_log_file = utils_misc.get_path(test.debugdir, iotests_log_file)
+            file(iotests_log_file, "w+").write(output)
```
Comment (xutian) [inline qemu/tests/rh_qemu_iotests.py]: bad code

File: qemu/tests/rh_qemu_iotests.py [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+            raise exceptions.TestFail(msg % (match.group(2), match.group(1)))
+
+    rpmbuild_dir = params.get("rpmbuild_dir", "/root/rpmbuild")
+    error_context.context("Install packages", logging.info)
+    utils_misc.yum_install(params.objects("configure_deps_pkgs"))
+    cwd = os.getcwd()
+    error_context.context("Get qemu source code", logging.info)
+    qemu_src_rpm_url = get_qemu_src_rpm_url()
+    if os.path.exists(rpmbuild_dir):
+        shutil.move(rpmbuild_dir, "%s_bak" % rpmbuild_dir)
```
Comment (xutian) [inline qemu/tests/rh_qemu_iotests.py]: not need to backup all direct, just qemu-kvm only  is enough.

File: qemu/tests/rh_qemu_iotests.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+    try:
+        qemu_src_dir = get_qemu_src_from_rpm(rpmbuild_dir, qemu_src_rpm_url)
+        run_configure_script(qemu_src_dir)
+        run_qemu_iotests(qemu_src_dir)
+    finally:
+        try:
+            os.chdir(cwd)
+            shutil.rmtree(rpmbuild_dir)
```
Comment (xutian) [inline qemu/tests/rh_qemu_iotests.py]: same

File: qemu/tests/rh_qemu_iotests.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+        query_format = params["query_format"]
+        download_rpm_cmd = params["download_rpm_cmd"]
+        get_src_cmd = params["get_src_cmd"]
+        qemu_spec = params.get("qemu_spec", 'SPECS/qemu-kvm.spec')
+        get_rpm_name_cmd = ("rpm -qf %s --queryformat=%s" %
+                            (utils_misc.get_qemu_binary(params), query_format))
+        src_rpm_name = process.system_output(get_rpm_name_cmd, shell=True)
+        retry_command(download_rpm_cmd % src_rpm_name)
+        spec = os.path.join(rpmbuild_dir, qemu_spec)
+        BUILD_dir = os.path.join(rpmbuild_dir, 'BUILD')
```
Comment (xutian) [inline qemu/tests/rh_qemu_iotests.py]: build_dir is ok

File: qemu/tests/rh_qemu_iotests.py [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+        cmd = get_src_cmd % (src_rpm_name, spec)
+        process.system(cmd, shell=True)
+        src_dir = os.listdir(BUILD_dir)[0]
```
Comment (xutian) [inline qemu/tests/rh_qemu_iotests.py]: same!!

File: qemu/tests/rh_qemu_iotests.py [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+    (qemu_src_dir, spec) = get_qemu_src(rpmbuild_dir)
+    try:
+        run_configure_script(qemu_src_dir)
+        run_qemu_iotests(qemu_src_dir)
+    finally:
+        try:
+            os.chdir(cwd)
+            process.system(rpmbuild_clean_cmd % spec, shell=True)
+        except Exception:
+            logging.error("Fail to clean test environment")
```
Comment (xutian) [inline qemu/tests/rh_qemu_iotests.py]: logging.warn is ok

File: qemu/tests/rh_qemu_iotests.py [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+            retry -= 1
+            try:
+                return process.system(cmd, shell=True)
+            except process.CmdError, detail:
+                msg = "Fail to execute command"
+                logging.error("%s: %s." % (msg, detail))
+        raise exceptions.TestError("%s after %s times retry: %s" %
+                                   (msg, max_retry, detail))
+
+    def get_qemu_src(rpmbuild_dir):
```
Comment (xutian) [inline qemu/tests/rh_qemu_iotests.py]: install_test
config_test
run_test

File: qemu/tests/rh_qemu_iotests.py [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+        os.environ["QEMU_IO_PROG"] = utils_misc.get_qemu_io_binary(params)
+        os.chdir(os.path.join(qemu_src_dir, iotests_dir))
+        cmd = './check'
+        if extra_options:
+            cmd += " %s" % extra_options
+        cmd += " -%s" % image_format
+        output = process.system_output(cmd, ignore_status=True, shell=True)
+        match = re.search(result_pattern, output, re.I | re.M)
+        if match:
+            iotests_log_file = "qemu_iotests_%s.log" % image_format
```
Comment (xutian) [inline qemu/tests/rh_qemu_iotests.py]: with open(iotests_log_file, 'w+') as log:
      log.write(output)
      log.flush()

File: qemu/tests/rh_qemu_iotests.py [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+            iotests_log_file = utils_misc.get_path(test.debugdir, iotests_log_file)
+            logfile = open(iotests_log_file, "w+")
+            logfile.write(output)
+            logfile.close()
+            msg = "Total test %s cases, %s failed"
+            raise exceptions.TestFail(msg % (match.group(2), match.group(1)))
+
+    rpmbuild_dir = params.get("rpmbuild_dir", "/root/rpmbuild")
```
Comment (xutian) [inline qemu/tests/rh_qemu_iotests.py]: build_root

**Comment by xutian** [comment]
Looks OK to me

**Pull request merged**